Add GDALG (GDAL Streamed Algorithm Format) driver (#11962)
OSGeo/gdal: doc/source/drivers/raster/gdalg.rst

.. _raster.gdalg:

================================================================================
GDALG: GDAL Streamed Algorithm
================================================================================

.. versionadded:: 3.11

.. shortname:: GDALG

.. built_in_by_default::

> This is a read-only driver that reads a JSON file containing a command line
> of the :ref:`gdal command line interface <gdal_program>`, that results in a
> on-the-fly / streamed raster dataset when this file is used as a input dataset
> for :cpp:func:`GDALOpenEx`, or anywhere in GDAL API or command line utilities
> where a raster input dataset is expected. GDALG files are conceptually close
> to :ref:`VRT (Virtual) files <raster.vrt>`, although the implementation is
> substantially different.

Reviewer: ```suggestion
This is a read-only driver that reads a JSON file containing an invocation
of the :ref:`gdal command line interface <gdal_program>`, that results in a
on-the-fly / streamed raster dataset when used as an input
to :cpp:func:`GDALOpenEx`, or anywhere else in GDAL
where a raster input dataset is expected. GDALG files are conceptually close
to :ref:`VRT (Virtual) files <raster.vrt>`, although the implementation is
substantially different.
```

The subset of commands of :program:`gdal` that can generate such streamed datasets

> indicate ``GDALG`` as a valid output format. This includes :ref:`gdal_raster_pipeline_subcommand`,

Reviewer: ```suggestion
indicate support for writing ``GDALG`` in the ``--help`` description for ``--output-format``. This includes :ref:`gdal_raster_pipeline_subcommand`,
```

Author: fixed

> indicates ``GDALG`` is a valid output format. This includes :ref:`gdal_raster_pipeline_subcommand`,

Reviewer: ```suggestion
indicate ``GDALG`` as a valid output format. This includes :ref:`gdal_raster_pipeline_subcommand`,
```

Author: fixed

> indicates ``GDALG`` is a valid output format. This includes :ref:`gdal_raster_pipeline_subcommand`,

Reviewer: How do they indicate this? By `gdal raster mosaic --formats` ?

Author: Their --help message shows:
```
-f, --of, --format, --output-format <OUTPUT-FORMAT> Output format ("GDALG" allowed)
```
Their documentation page also has a paragraph specific to GDALG output (I've edited the gdal raster mosaic and gdal raster stack pages which lacked those updates)

:ref:`gdal_raster_mosaic_subcommand` or :ref:`gdal_raster_stack_subcommand`.

> It is recommended that GDALG files use the ``.gdalg`` extension.

Reviewer: I'm guessing this is to facilitate detection? (I think `.gdalg.json` would be friendlier for users trying to edit these files)

Reviewer: Yes, it's my experience hand-editing `.vrt` files that makes me wonder about `gdalalg.json` :)

Author: extension changed to ``.gdalg.json``

The JSON document must include a ``type`` member with the value ``gdal_streamed_alg``,
and a ``command_line`` member, which is very close to the one that would be used
to generate a materialized dataset, but using the ``stream`` output format.

.. code-block:: json

    {
        "type": "gdal_streamed_alg",

>         "command_line": "gdal raster mosaic in1.tif in2.tif --output-format stream streamed_dataset"

Reviewer: Did you decide against "Variant B" in https://github.com/OSGeo/gdal/issues/11837, or is that to follow?

Author: Neither against, nor to immediately to follow. I made the choice of simplicity for now. Variant B can be implemented at a later stage if we think it is valuable to have.
